server/db/mysql/adapter.go

```diff
@@ -867,47 +921,68 @@ func (a *adapter) UserDelete(uid t.Uid, hard bool) error {
-
 		// Disable all subscriptions to topics where the user is the owner.
 		if _, err = tx.Exec("UPDATE subscriptions LEFT JOIN topics ON subscriptions.topic=topics.name "+
-			"SET subscriptions.updatedAt=?, subscriptions.deletedAt=? WHERE topics.owner=?",
+			"SET subscriptions.updatedat=?, subscriptions.deletedat=? WHERE topics.owner=?",
 			now, now, decoded_uid); err != nil {
 			return err
 		}
 		// Disable all topics where the user is the owner.
-		if _, err = tx.Exec("UPDATE topics SET updatedAt=?, deletedAt=? WHERE owner=?",
+		if _, err = tx.Exec("UPDATE topics SET updatedat=?, state=?, stateat=? WHERE owner=?",
+			now, t.StateDeleted, now, decoded_uid); err != nil {
+			return err
+		}
+		// Disable p2p topics with the user (p2p topic's owner is 0)
+		if _, err = tx.Exec("UPDATE topics LEFT JOIN subscriptions ON topics.name=subscriptions.topic "+
+			"SET topics.updatedat=?, topics.state=?, topics.stateat=? WHERE topics.owner=0 AND subscriptions.userid=?",
+			now, t.StateDeleted, now, decoded_uid); err != nil {
+			return err
+		}
+
+		// Disable the other user's subscription to a disabled p2p topic.
+		if _, err = tx.Exec("UPDATE subscriptions AS s_one LEFT JOIN subscriptions AS s_two "+
+			"ON s_one.topic=s_two.topic "+
+			"SET s_two.updatedat=?, s_two.deletedat=? WHERE s_one.userid=?",
 			now, now, decoded_uid); err != nil {
 			return err
 		}
 
 		// Disable user.
-		if _, err = tx.Exec("UPDATE users SET updatedAt=?, deletedAt=? WHERE id=?", now, now, decoded_uid); err != nil {
+		if _, err = tx.Exec("UPDATE users SET updatedat=?, state=?, stateat=? WHERE id=?",
+			now, t.StateDeleted, now, decoded_uid); err != nil {
 			return err
 		}
 	}
 
 	return tx.Commit()
 }
 
-func (a *adapter) UserGetDisabled(since time.Time) ([]t.Uid, error) {
-	rows, err := a.db.Queryx("SELECT id FROM users WHERE deletedat>=?", since)
-	if err != nil {
-		return nil, err
+// topicStateForUser is called by UserUpdate when the update contains state change.
+func (a *adapter) topicStateForUser(tx *sqlx.Tx, decoded_uid int64, now time.Time, update interface{}) error {
+	var err error
+
+	state, ok := update.(t.ObjState)
+	if !ok {
+		return t.ErrMalformed
 	}
 
-	var uids []t.Uid
-	for rows.Next() {
-		var userId int64
-		if err = rows.Scan(&userId); err != nil {
-			uids = nil
-			break
-		}
-		uids = append(uids, store.EncodeUid(userId))
+	if now.IsZero() {
+		now = t.TimeNow()
+	}
+
+	// Change state of all topics where the user is the owner.
+	if _, err = tx.Exec("UPDATE topics SET state=?, stateat=? WHERE owner=?", state, now, decoded_uid); err != nil {
+		return err
 	}
-	rows.Close()
 
-	return uids, err
+	// Change state of p2p topics with the user (p2p topic's owner is 0)
+	if _, err = tx.Exec("UPDATE topics LEFT JOIN subscriptions ON topics.name=subscriptions.topic "+
+		"SET topics.state=?, topics.stateat=? WHERE topics.owner=0 AND subscriptions.userid=?",
+		state, now, decoded_uid); err != nil {
+		return err
+	}
+
```

Comment:
Should we update subscriptions here as well?

Reply:
Added a comment that they do not need to be disabled. But that reminded me that I should disable ability to create new subscriptions for suspended users.